멘토링 세션 생성 및 등록 구현

1seyoung opened the pull request:
### 구현 내용

디비 테이블 추가
- 멘토링 세션 테이블(mentoring_sessions)
- 세션 참가자 테이블(session_participants) 

기능 구현
- 관리자 권한을 가진 사용자의 멘토링 세션 생성 기능
- 학생의 멘토링 세션 참가 신청 기능 (선착순)
- 비관적 락(SELECT FOR UPDATE)을 사용

미구현 사항
- [x] 멘토링 등록 취소 기능
- [ ] 취소 시 대기 인원 등록
- [ ] 기능 테스트 진행 x (시간이 없어 일단 올려봅니다)

### 메인 리뷰어 지정 및 Due Date
@blue000927

### 관련 이슈
#23 
#24 

체크리스트

- [x]  PR 제목을 명령형으로 작성했습니다.
- [x]  PR을 연관되는 github issue에 연결했습니다.
- [ ]  리뷰 리퀘스트 전에 셀프 리뷰를 진행했습니다.

1seyoung committed:
feat : 멘토링 세션 생성 기능 추가

1seyoung committed:
feat : 멘토링 세션 생성 기능 추가

1seyoung committed:
Copyfeat: 멘토링 세션 기능 추가
- 멘토링 세션 생성 기능 구현
- 학생의 멘토링 세션 등록 기능 구현
- 동시성 처리를 위한 비관적 락(FOR UPDATE) 적용
- 멘토링 세션 관련 도메인 객체 및 매퍼 구현

1seyoung committed:
Merge remote-tracking branch 'origin/feature/mentoring-session' into feature/mentoring-session

# Conflicts:
#	src/main/java/com/gatheria/controller/MentoringSessionController.java
#	src/main/java/com/gatheria/domain/MentoringSession.java
#	src/main/java/com/gatheria/mapper/MentoringSessionMapper.java
#	src/main/java/com/gatheria/service/MentoringSessionService.java

1seyoung committed:
fix : 동시성 코드 삭제

blue000927 commented on src/main/java/com/gatheria/controller/MentoringSessionController.java:
```diff
@@ -0,0 +1,57 @@
+    this.mentoringSessionService = mentoringSessionService;
+  }
+
+  @PostMapping("/create")
+  public ResponseEntity<MentoringSessionCreateResponseDto> createMentoringSession(
```
POST가 리소스를 생성한다는 의미이므로 "/create"는 제거하셔도 됩니다.

blue000927 commented on src/main/java/com/gatheria/controller/MentoringSessionController.java:
```diff
@@ -0,0 +1,57 @@
+
+  private final MentoringSessionService mentoringSessionService;
+
+  private static final String ADMIN_KEY = "admin123";
```
이렇게 관리하지 말고 어드민용 테이블을 추가하는 것이 좋지 않을까요? 혹은 유저 테이블에 어드민 데이터를 넣거나..

1seyoung replied:
Admin 생성 및 로그인 기능을 추가했고, 로직을 분리했습니다.  이 외에 Admin 여부를 확인하고 기능을 실행하는 로직은 다른 PR에서 추가하겠습니다.

blue000927 commented on src/main/java/com/gatheria/domain/MentoringSession.java:
```diff
@@ -0,0 +1,49 @@
+    this.sessionDate = sessionDate;
+    this.maxParticipants = maxParticipants;
+    this.currentParticipants = 0;
+    this.status = "OPEN";
```
enum 활용 부탁드립니다.

1seyoung replied:
전체 상태 관련 코드 enum 적용했습니다

blue000927 commented on src/main/java/com/gatheria/domain/SessionParticipant.java:
```diff
@@ -0,0 +1,40 @@
+  private Long id;
+  private final Long sessionId;
+  private final Long studentId;
+  private LocalDateTime registrationTime; // final 제거
```
주석을 제거해도 될거같아요.

blue000927 commented on src/main/java/com/gatheria/domain/SessionParticipant.java:
```diff
@@ -0,0 +1,40 @@
+    this.sessionId = sessionId;
+    this.studentId = studentId;
+    this.registrationTime = LocalDateTime.now();
+    this.status = "REGISTERED";
```
ditto (enum)

blue000927 commented on src/main/java/com/gatheria/domain/SessionParticipant.java:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  public void cancel() {
+    this.status = "CANCELLED";
```
ditto (enum)

blue000927 commented on src/main/java/com/gatheria/domain/SessionParticipant.java:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  public static SessionParticipant reRegister(SessionParticipant existing) {
+    existing.status = "REGISTERED";
```
ditto (enum)

blue000927 commented on src/main/java/com/gatheria/dto/response/SessionRegistrationResponseDto.java:
```diff
@@ -0,0 +1,41 @@
+        sessionId,
+        sessionTitle,
+        sessionDate,
+        "멘토링 세션 등록 완료.",
```
메시지가 꼭 필요한가요?? Status Code로 200을 내려주면 성공이 아닌가해서요...!

1seyoung replied:
null 로 작성했는데 그냥 메시지 전달하는 것 자체가 의미가 없을까요?

blue000927 replied:
OK 응답을 보면 성공이다보니 의미가 아예 없진 않겠지만 큰 효과가 느껴지진 않을것같아서요 ㅎㅎ
세영님께서 프론트 코드도 작업하고 계시니 프론트 입장에서 생각해서 판단해주세요~

blue000927 commented on src/main/java/com/gatheria/service/MentoringSessionService.java:
```diff
@@ -0,0 +1,88 @@
+    SessionParticipant existing = mentoringSessionMapper.findBySessionAndStudent(sessionId,
+        authInfo.getMemberId());
+
+    if (existing != null && "REGISTERED".equals(existing.getStatus())) {
```
ditto (enum)

1seyoung committed:
feat: 관리자 회원가입 및 로그인 기능 구현
- Admin 회원가입 기능 추가
- 관리자 로그인 기능 추가
- Admin 엔티티

1seyoung committed:
feat: 멘토링 세션 상태 및 대기 등록 기능 개선
- 멘토링 세션 상태를 Enum 으로 관리
- 멘토링 참가 등록 상태 Enum 으로 관리
- 대기 신청 시작/마감 시간 (`waitingStartDate`, `waitingEndDate`) 추가

blue000927 commented on src/main/java/com/gatheria/domain/type/MentoringStatus.java:
```diff
@@ -0,0 +1,12 @@
+@Getter
+public enum MentoringStatus {
+  SCHEDULED,
+  WAITING_OPEN,
+  WAITING_CLOSED,
```
우리 동시성 하나만 해도 빡세다보니 대기열 처리는 우선 제거하고, 추후 이슈를 파서 진행해 보면 어떨까요?

1seyoung replied:
SCHEDULED,  일정 잡힌 상태
  WAITING_OPEN, 등록 오픈 상태
  WAITING_CLOSED, 등록 마감 상태
  COMPLETED, 세션 종료
  CANCELLED 세션 취소
  
  를 나타내게 했는데 대기열은 나중에 구현하더라도 해당 부분은 있어도 괜찮지않을까요?

pjy1368 replied:
아~ 좋습니다. enum은 유지합시다

1seyoung committed:
Merge branch 'main' into feature/mentoring-session
